File: framework/languages/java/samples/java/Bingo/Server/Api/src/main/java/systems/zlink/samples/bingo/server/api/handlers/MatchBingoHandler.java
package systems.zlink.samples.bingo.server.api.handlers;

import systems.zlink.framework.channels.ZLinkRouteClient;
import systems.zlink.framework.ZLinkMessageContext;
import systems.zlink.framework.channels.ZLinkRequestHandler;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.samples.bingo.server.configuration.SampleNames;
import systems.zlink.samples.bingo.server.configuration.SampleTimings;
import systems.zlink.samples.bingo.shared.contracts.BingoMessages;
import systems.zlink.samples.bingo.shared.contracts.Messages;

@ZLinkHandlerGroup(SampleNames.ApiChannel)
public final class MatchBingoHandler
    implements ZLinkRequestHandler<
        Messages.MatchBingoApiReq,
        Messages.MatchBingoApiRes> {
    private final ZLinkRouteClient routes;

    public MatchBingoHandler(ZLinkRouteClient routes) {
        this.routes = routes;
    }

    @Override
    public java.util.concurrent.CompletionStage<Messages.MatchBingoApiRes> handle(
        Messages.MatchBingoApiReq request,
        ZLinkMessageContext context) {
        return routes.requestToChannel(
                SampleNames.RoomSpotDiscovery,
                BingoMessages.allocateBingoRoomReq(
                    request.getMode(),
                    request.getActorId()))
            .timeout(SampleTimings.RequestTimeout)
            .submit(Messages.AllocateBingoRoomRes.class)
            .thenApply(allocated -> BingoMessages.matchBingoApiRes(allocated.getRoomId()));
    }
}

File: framework/languages/java/samples/kotlin/Bingo/Server/Api/src/main/kotlin/systems/zlink/samples/kotlin/bingo/server/api/handlers/MatchBingoHandler.kt
package systems.zlink.samples.kotlin.bingo.server.api.handlers

import kotlinx.coroutines.future.await
import systems.zlink.framework.channels.ZLinkRouteClient
import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.kotlin.ZLinkSuspendingRequestHandler
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.samples.kotlin.bingo.server.configuration.SampleNames
import systems.zlink.samples.kotlin.bingo.server.configuration.SampleTimings
import systems.zlink.samples.kotlin.bingo.shared.contracts.AllocateBingoRoomReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.AllocateBingoRoomRes
import systems.zlink.samples.kotlin.bingo.shared.contracts.MatchBingoApiReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.MatchBingoApiRes

@ZLinkHandlerGroup(SampleNames.ApiChannel)
class MatchBingoHandler(
    private val routes: ZLinkRouteClient,
) : ZLinkSuspendingRequestHandler<MatchBingoApiReq, MatchBingoApiRes> {
    override suspend fun handle(
        request: MatchBingoApiReq,
        context: ZLinkMessageContext,
    ) = run {
        val allocated = routes
            .requestToChannel(
                SampleNames.RoomSpotDiscovery,
                AllocateBingoRoomReq(
                    request.actorId,
                    request.mode,
                ),
            )
            .timeout(SampleTimings.RequestTimeout)
            .submit(AllocateBingoRoomRes::class.java)
            .await()

        MatchBingoApiRes(allocated.roomId)
    }
}
